fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1091,25 +1091,210 @@ inline bool tenantMapChanging(MutationRef const& mutation) {
+	ASSERT_EQ(result.size(), 1);
+	ASSERT(result.front().param1 == param1);
+	ASSERT(result.front().param2 == param2);
+
+	// multiple tenant
+	int64_t tid1 = deterministicRandom()->randomInt64(0, mapSize - 1);
+	int64_t tid2 = deterministicRandom()->randomInt64(tid1 + 1, mapSize) * 2;
+	tid1 *= 2;
+	KeyRef prefix1 = TenantAPI::idToPrefix(arena, tid1);
+	param1 = deterministicRandom()->coinflip() ? prefix1 : prefix1.withSuffix("a"_sr, arena); // align or not
+	KeyRef prefix2 = TenantAPI::idToPrefix(arena, tid2);
+	bool tailAligned = deterministicRandom()->coinflip();
+	param2 = tailAligned ? prefix2 : prefix2.withSuffix("a"_sr, arena);
+	int targetSize = (tid2 - tid1) / 2 + (!tailAligned);
+	mutation.param1 = param1;
+	mutation.param2 = param2;
+	result = splitClearRangeByTenant(arena, mutation, tenantMap);
+	ASSERT_EQ(result.size(), targetSize);
+	ASSERT(result.front().param1 == param1);
+	if (tailAligned) {
+		KeyRange r = prefixRange(TenantAPI::idToPrefix(tid2 - 2));
+		ASSERT(r == KeyRangeRef(result.back().param1, result.back().param2));
+	} else {
+		ASSERT(result.back().param1 == prefix2);
+		ASSERT(result.back().param2 == param2);
+	}
+
+	return Void();
+}
+
+void replaceRawClearRanges(Arena& arena,
+                           VectorRef<MutationRef>& mutations,
+                           std::vector<std::vector<MutationRef>>& splitMutations,
+                           size_t totalSize) {
+	if (splitMutations.empty())
+		return;
+	mutations.resize(arena, totalSize);
+	// place from back
+	int curr = totalSize - 1;
+	bool doValidation = EXPENSIVE_VALIDATION;
+	for (int i = mutations.size() - 1; i >= 0; --i) {
+		if (splitMutations.empty() && !doValidation)
+			break;
+
+		if (splitMutations.empty()) {
+			ASSERT_EQ(curr, i);
+			curr--;
+			continue;
+		}
+
+		if (mutations[i].type == MutationRef::ClearRange) {
+			// replace with tenant aligned mutations
+			auto& currMutations = splitMutations.back();
+			while (!currMutations.empty()) {
+				mutations[curr] = currMutations.back();
+				currMutations.pop_back();
+				curr--;
+			}
+			splitMutations.pop_back();
+		} else {
+			ASSERT_GE(curr, i);
+			if (curr > i) {
+				mutations[curr] = mutations[i];
+			}
+			curr--;
+		}
+	}
+
+	ASSERT_EQ(splitMutations.size(), 0);
+}
+
 // Return success and properly split clear range mutations if all tenant check pass. Otherwise, return corresponding
 // error
-Error validateAndProcessTenantAccess(VectorRef<MutationRef> mutations,
+Error validateAndProcessTenantAccess(Arena& arena,
+                                     VectorRef<MutationRef>& mutations,
                                      ProxyCommitData* const pProxyCommitData,
                                      Optional<UID> debugId = Optional<UID>(),
                                      const char* context = "") {
 	bool changeTenant = false;
 	bool writeNormalKey = false;
 
+	std::vector<std::vector<MutationRef>> splitMutations;
+	size_t newMutationSize = mutations.size();
```

Comment:
Another option here is to iterate `mutations` by index and directly store the split mutations into the existing vector. This can be done as follows:

1. If the clear range isn't changing, do nothing
2. If the clear range is being split, store the first updated mutation at the original index
3. Store subsequent mutations at the end of the vector

This could allow you to avoid extra scans and temporary data structures

Reply:
Isn't the operation order important?

Reply:
Yeah, you're right. It's still possible to do this in a single forward pass by copying to a new vector (only if needed), but unlike your solution it's probably harder to avoid copying the elements before the first split. I guess more accurately, we can choose between having to do two passes through the elements before the first split or after, since the first copy in my suggestion will require a second pass through the first elements and your approach requires a second pass through the remaining elements in reverse.

Because this situation is likely to be quite rare and non-critical, I'm inclined to say we should implement it the way that is the most straightforward to understand. My initial thought is that a direct copy to a new vector once a split is detected would be easier, but I could be wrong about that.